jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/ProjectCommand.java

```diff
@@ -79,4 +107,73 @@ private static String getWorkspaceInvisibleProjectName(IPath workspacePath) {
         String fileName = workspacePath.toFile().getName();
         return fileName + "_" + Integer.toHexString(workspacePath.toPortableString().hashCode());
     }
+
+    public static List<MainClass> getMainMethod(IProgressMonitor monitor) {
+        final List<MainClass> res = new ArrayList<>();
+        IJavaSearchScope scope = SearchEngine.createWorkspaceScope();
+        SearchPattern pattern = SearchPattern.createPattern("main(String[]) void", IJavaSearchConstants.METHOD,
+                IJavaSearchConstants.DECLARATIONS, SearchPattern.R_EXACT_MATCH | SearchPattern.R_CASE_SENSITIVE);
+        SearchRequestor requestor = new SearchRequestor() {
+            @Override
+            public void acceptSearchMatch(SearchMatch match) {
+                Object element = match.getElement();
+                if (element instanceof IMethod) {
+                    IMethod method = (IMethod) element;
+                    try {
+                        if (method.isMainMethod()) {
+                            IResource resource = method.getResource();
+                            if (resource == null) {
+                                return;
+                            }
+                            IProject project = resource.getProject();
+                            if (project == null) {
+                                return;
+                            }
+                            String mainClass = method.getDeclaringType().getFullyQualifiedName();
+                            IJavaProject javaProject = method.getJavaProject();
+                            if (javaProject != null) {
+                                String moduleName = getModuleName(javaProject);
+                                if (moduleName != null) {
+                                    mainClass = moduleName + "/" + mainClass;
+                                }
+                            }
+                            String filePath = null;
+                            if (match.getResource() instanceof IFile) {
+                                try {
+                                    filePath = match.getResource().getLocation().toOSString();
+                                } catch (Exception ex) {
```

Comment:
Does `match.getResource().getLocation().toOSString()` throw the general typed Exception?